Go: Promote non-httponly cookie query, and add insecure cookie query

joefarebrother opened the pull request:
Promotes `go/cookie-httponly-not-set` from experimental, and adds `go/cookie-secure-not-set` query.

owen-mc commented on go/ql/lib/semmle/go/concepts/HTTP.qll:
```diff
@@ -380,4 +380,96 @@ module Http {
     predicate guardedBy(DataFlow::Node check) { super.guardedBy(check) }
   }
+
+  /** Provides a class for modeling HTTP response cookie writes. */
```
```suggestion
  /** Provides a class for modeling new HTTP response cookie write APIs. */
```

owen-mc commented on go/ql/lib/semmle/go/concepts/HTTP.qll:
```diff
@@ -380,4 +380,96 @@ module Http {
+    DataFlow::Node getHttpOnly() { result = super.getHttpOnly() }
+  }
+
+  /** Provides a class for modeling the options of an HTTP cookie. */
```
```suggestion
  /** Provides a class for modeling new APIs for the options of an HTTP cookie. */
```

owen-mc commented on go/ql/lib/semmle/go/frameworks/stdlib/NetHttp.qll:
```diff
@@ -293,4 +293,38 @@ module NetHttp {
+
+  private class CookieFieldWrite extends Http::CookieOptions::Range {
+    Write w;
+    Field f;
```
Rather than storing these as fields, they should be introduced in the characteristic predicate using `exists`, since they aren't referenced by any of the member predicates.

joefarebrother commented:
Addressed comments 
I can re-add the `gorilla/sessions` models; they were initially ommited due to a slight complication in which there isn't a sensitive name expression to alert for as the source; however it can be handled by alerting for the cookie constructor with a name of something like "Gorilla session".

owen-mc commented on go/ql/lib/semmle/go/security/CookieWithoutHttpOnly.qll:
```diff
@@ -0,0 +1,77 @@
+
+/** Holds if `cw` has the `HttpOnly` attribute left at its default value of `false`. */
+predicate isNonHttpOnlyDefault(Http::CookieWrite cw) {
+  not BooleanCookieHttpOnlyFlow::flow(_, cw.getHttpOnly())
```
There's actually a predicate for this exact thing.
```suggestion
  not BooleanCookieHttpOnlyFlow::flowTo(cw.getHttpOnly())
```

owen-mc commented on go/ql/test/query-tests/Security/CWE-1004/vendor/github.com/gorilla/sessions/stub.go:
Delete this file

owen-mc commented on go/ql/test/query-tests/Security/CWE-1004/go.mod:
Remove mentions of gorilla from this file.

owen-mc commented on go/ql/src/Security/CWE-1004/CookieWithoutHttpOnly.ql:
```diff
@@ -0,0 +1,25 @@
+  Http::CookieWrite cw, string name, SensitiveCookieNameFlow::PathNode source,
+  SensitiveCookieNameFlow::PathNode sink
+where
+  isSensitiveCookie(cw, name, source, sink) and
```
Oops, I missed that somehow.

joefarebrother commented:
Autofix results look good.
